enhancements/cluster-constraints-properties.md

```diff
@@ -0,0 +1,172 @@
+
+#### Cluster properties
+
+As a cluster admin, I would like to provide cluster properties that are relevant to cluster availability and capabilities so that operator authors can depend on those properties if needed to ensure their operators are installable and functioning properly in the cluster.
+
+#### Bundle properties and constraints
+
+As an operator author, I would like to include constraints that rely on cluster properties so my operators in my bundles so that they can only be installed on suitable clusters.
+
+I also would like to provide essential properties that are used in cluster constraints in my bundle so that they can be screened at runtime if the clusters have certain requirements that may impact my operators' operation.
+
+### Design Details
+
+#### Cluster information API
+
+The cluster constraint API is constructed using CustomResourceDefinition (CRD):
+
+```yaml=
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: constraints.operators.coreos.com
+spec:
+  group: operators.coreos.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                constraints:
+                  type: string
+                properties:
+                  type: string
+  scope: Cluster
+  names:
+    plural: constraints
+    singular: constraint
+    kind: constraint
+    shortNames:
+    - constraint
+```
+
+The CRD is cluster-scoped as the CR(s) are representing cluster information. The cluster administrators have RBAC permission to create CustomResource (CR) that will contain cluster constraints and properties. For example:
+
+```yaml=
+apiVersion: operators.coreos.com/v2
+kind: Constraint
+metadata:
+  name: cluster-constraints
+spec:
+  constraints: '[{"type":"olm.constraint","value":{"evaluator":{"id":"cel"},"rule":"properties.exists(p, p.type == \"certified\")","message":"require to have certified property","action":{"id":"require"}}}]'
+  properties: '[{"type":"olm.k8s","value":{"version":"1.21"}},{"type":"olm.label","value":{"label":"secured"}}]'
+```
+
+The CR's spec contains two fields: `constraints` and `properties`. The `constraints` field is a list of `olm.constraint` type (See [1](https://github.com/operator-framework/enhancements/pull/91)). Each constraint is specified using the following syntax:
+
+```json=
+{
+   "type":"olm.constraint",
+   "value":{
+      "evaluator":{
+         "id":"cel"
+      },
+      "rule": "properties.exists(p, p.type == \"certified\")",
+      "message": "require to have certified property",
+      "action":{
+         "id":"require"
+      }
+   }
+}
+```
+
+The `properties` field is a list of properties that are specific to the cluster at the runtime. Each property is specified using the following syntax:
```

Comment:
Will be able to work with ranges? If yes, could add an example? 
Also, what happens when is not possible to install because of the constrain? could we:
a) Ensure that for any predefined type we will have a good and nice msg to output in order to make it easier for the consumer to know why the operator was not installed? 
b) If I properly understood, in the EP: https://github.com/operator-framework/enhancements/pull/91 we are able to create generic rules. What message we will show in this case? WDYT about the operator author also providing the msg?